diff --color -ruN gcc-rustc/gcc/config/sh/sh-c.cc gcc-rustc-kos/gcc/config/sh/sh-c.cc
--- gcc-rustc/gcc/config/sh/sh-c.cc	2024-02-02 23:52:38.294387173 -0600
+++ gcc-rustc-kos/gcc/config/sh/sh-c.cc	2024-02-02 23:52:57.783480557 -0600
@@ -141,4 +141,11 @@
 
   cpp_define_formatted (pfile, "__SH_ATOMIC_MODEL_%s__",
 			selected_atomic_model ().cdef_name);
+
+  /* Custom built-in defines for KallistiOS */
+  builtin_define ("__KOS_GCC_PATCHED__");
+  cpp_define_formatted (pfile, "__KOS_GCC_PATCHLEVEL__=%d",
+			2023010200);
+  /* Toolchain supports setting up stack for 32MB */
+  builtin_define ("__KOS_GCC_32MB__");
 }
diff --color -ruN gcc-rustc/gcc/configure gcc-rustc-kos/gcc/configure
--- gcc-rustc/gcc/configure	2024-02-02 23:52:38.301387207 -0600
+++ gcc-rustc-kos/gcc/configure	2024-02-02 23:52:57.786480571 -0600
@@ -13108,7 +13108,7 @@
     target_thread_file='single'
     ;;
   aix | dce | lynx | mipssde | posix | rtems | \
-  single | tpf | vxworks | win32 | mcf)
+  single | tpf | vxworks | win32 | kos | mcf)
     target_thread_file=${enable_threads}
     ;;
   *)
diff --color -ruN gcc-rustc/libgcc/config/sh/t-sh gcc-rustc-kos/libgcc/config/sh/t-sh
--- gcc-rustc/libgcc/config/sh/t-sh	2024-02-02 23:52:41.142400820 -0600
+++ gcc-rustc-kos/libgcc/config/sh/t-sh	2024-02-02 23:52:57.787480576 -0600
@@ -23,6 +23,8 @@
   $(LIB1ASMFUNCS_CACHE)
 LIB1ASMFUNCS_CACHE = _ic_invalidate _ic_invalidate_array
 
+LIB2ADD = $(srcdir)/config/sh/fake-kos.S
+
 crt1.o: $(srcdir)/config/sh/crt1.S
 	$(gcc_compile) -c $<
 
diff --color -ruN gcc-rustc/libgcc/configure gcc-rustc-kos/libgcc/configure
--- gcc-rustc/libgcc/configure	2024-02-02 23:52:41.146400839 -0600
+++ gcc-rustc-kos/libgcc/configure	2024-02-02 23:52:57.787480576 -0600
@@ -5732,6 +5732,7 @@
     tpf)	thread_header=config/s390/gthr-tpf.h ;;
     vxworks)	thread_header=config/gthr-vxworks.h ;;
     win32)	thread_header=config/i386/gthr-win32.h ;;
+    kos)	thread_header=config/sh/gthr-kos.h ;;
     mcf)	thread_header=config/i386/gthr-mcf.h ;;
 esac
 
diff --color -ruN gcc-rustc/libobjc/configure gcc-rustc-kos/libobjc/configure
--- gcc-rustc/libobjc/configure	2024-02-02 23:52:41.498402525 -0600
+++ gcc-rustc-kos/libobjc/configure	2024-02-02 23:52:57.788480581 -0600
@@ -2924,11 +2924,9 @@
 
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
-#include <stdio.h>
 int
 main ()
 {
-printf ("hello world\n");
   ;
   return 0;
 }
diff --color -ruN gcc-rustc/libobjc/Makefile.in gcc-rustc-kos/libobjc/Makefile.in
--- gcc-rustc/libobjc/Makefile.in	2024-02-02 23:52:41.497402521 -0600
+++ gcc-rustc-kos/libobjc/Makefile.in	2024-02-02 23:52:57.788480581 -0600
@@ -308,14 +308,16 @@
 $(srcdir)/aclocal.m4: @MAINT@ $(aclocal_deps)
 	cd $(srcdir) && $(ACLOCAL) $(ACLOCAL_AMFLAGS)
 
-install: install-libs install-headers
+install-strip: INSTALL_STRIP_FLAG = -s
+install install-strip: install-libs install-headers
 
 install-libs: installdirs
 	$(SHELL) $(multi_basedir)/mkinstalldirs $(DESTDIR)$(toolexeclibdir)
-	$(LIBTOOL_INSTALL) $(INSTALL) libobjc$(libsuffix).la $(DESTDIR)$(toolexeclibdir);
+	$(LIBTOOL_INSTALL) $(INSTALL) $(INSTALL_STRIP_FLAG) \
+	  libobjc$(libsuffix).la $(DESTDIR)$(toolexeclibdir);
 	if [ "$(OBJC_BOEHM_GC)" ]; then \
-	  $(LIBTOOL_INSTALL) $(INSTALL) libobjc_gc$(libsuffix).la \
-				$(DESTDIR)$(toolexeclibdir);\
+	  $(LIBTOOL_INSTALL) $(INSTALL) $(INSTALL_STRIP_FLAG) \
+	    libobjc_gc$(libsuffix).la $(DESTDIR)$(toolexeclibdir);\
 	fi
 	$(MULTIDO) $(FLAGS_TO_PASS) multi-do DO="$@"
 	@-$(LIBTOOL) --mode=finish $(DESTDIR)$(toolexeclibdir)
@@ -328,7 +330,7 @@
 	  $(INSTALL_DATA) $${realfile} $(DESTDIR)$(libsubdir)/$(includedirname)/objc; \
 	done
 
-check uninstall install-strip dist installcheck installdirs:
+check uninstall dist installcheck installdirs:
 
 mostlyclean:
 	-$(LIBTOOL_CLEAN) rm -f libobjc$(libsuffix).la libobjc_gc$(libsuffix).la *.lo
diff --color -ruN gcc-rustc/libstdc++-v3/config/cpu/sh/atomicity.h gcc-rustc-kos/libstdc++-v3/config/cpu/sh/atomicity.h
--- gcc-rustc/libstdc++-v3/config/cpu/sh/atomicity.h	2024-02-02 23:52:41.633403172 -0600
+++ gcc-rustc-kos/libstdc++-v3/config/cpu/sh/atomicity.h	2024-02-02 23:52:57.788480581 -0600
@@ -22,14 +22,40 @@
 // see the files COPYING3 and COPYING.RUNTIME respectively.  If not, see
 // <http://www.gnu.org/licenses/>.
 
-// Use the default atomicity stuff, which will use __atomic* builtins
-// if threads are available, or the *_single functions on single-thread
-// configurations.
-// Actually we wouldn't need this header at all, but because of PR 53579
-// libstdc++'s configury will not pickup the -matomic-model= option when
-// set in the environment.  This makes it impossible to enable the proper
-// atomic model on SH without modifying GCC itself, because libstdc++ always
-// thinks the target doesn't do any atomics and uses the default mutex based
-// implementation from cpu/generic/atomicity_mutex.
+/* This is generic/atomicity.h */
 
 #include <ext/atomicity.h>
+#include <ext/concurrence.h>
+
+namespace 
+{
+  __gnu_cxx::__mutex&
+  get_atomic_mutex()
+  {
+    static __gnu_cxx::__mutex atomic_mutex;
+    return atomic_mutex;
+  }
+} // anonymous namespace
+
+namespace __gnu_cxx _GLIBCXX_VISIBILITY(default)
+{
+_GLIBCXX_BEGIN_NAMESPACE_VERSION
+
+  _Atomic_word
+  __attribute__ ((__unused__))
+  __exchange_and_add(volatile _Atomic_word* __mem, int __val) throw ()
+  {
+    __gnu_cxx::__scoped_lock sentry(get_atomic_mutex());
+    _Atomic_word __result;
+    __result = *__mem;
+    *__mem += __val;
+    return __result;
+  }
+
+  void
+  __attribute__ ((__unused__))
+  __atomic_add(volatile _Atomic_word* __mem, int __val) throw ()
+  { __exchange_and_add(__mem, __val); }
+
+_GLIBCXX_END_NAMESPACE_VERSION
+} // namespace
diff --color -ruN gcc-rustc/libstdc++-v3/configure gcc-rustc-kos/libstdc++-v3/configure
--- gcc-rustc/libstdc++-v3/configure	2024-02-02 23:52:41.639403201 -0600
+++ gcc-rustc-kos/libstdc++-v3/configure	2024-02-02 23:52:57.792480600 -0600
@@ -15959,6 +15959,7 @@
     tpf)	thread_header=config/s390/gthr-tpf.h ;;
     vxworks)	thread_header=config/gthr-vxworks.h ;;
     win32)	thread_header=config/i386/gthr-win32.h ;;
+    kos)	thread_header=config/sh/gthr-kos.h ;;
     mcf)	thread_header=config/i386/gthr-mcf.h ;;
 esac
 
