diff --color -ruN rustc_codegen_gcc/build_system/build_sysroot/Cargo.toml rustc_codegen_gcc-fixed/build_system/build_sysroot/Cargo.toml
--- rustc_codegen_gcc/build_system/build_sysroot/Cargo.toml	2024-04-10 12:55:03.365847139 -0500
+++ rustc_codegen_gcc-fixed/build_system/build_sysroot/Cargo.toml	2024-04-10 12:56:01.435157041 -0500
@@ -8,14 +8,14 @@
 core = { path = "./sysroot_src/library/core" }
 compiler_builtins = "0.1"
 alloc = { path = "./sysroot_src/library/alloc" }
-std = { path = "./sysroot_src/library/std", features = ["panic_unwind", "backtrace"] }
-test = { path = "./sysroot_src/library/test" }
+#std = { path = "./sysroot_src/library/std", features = ["panic_unwind", "backtrace"] }
+#test = { path = "./sysroot_src/library/test" }
 proc_macro = { path = "./sysroot_src/library/proc_macro" }
 
 [patch.crates-io]
 rustc-std-workspace-core = { path = "./sysroot_src/library/rustc-std-workspace-core" }
 rustc-std-workspace-alloc = { path = "./sysroot_src/library/rustc-std-workspace-alloc" }
-rustc-std-workspace-std = { path = "./sysroot_src/library/rustc-std-workspace-std" }
+#rustc-std-workspace-std = { path = "./sysroot_src/library/rustc-std-workspace-std" }
 
 [profile.release]
 debug = "limited"
