diff --color -ruN rustc_codegen_gcc/build_system/src/config.rs rustc_codegen_gcc-fixed/build_system/src/config.rs
--- rustc_codegen_gcc/build_system/src/config.rs	2024-04-10 12:55:03.365847139 -0500
+++ rustc_codegen_gcc-fixed/build_system/src/config.rs	2024-04-10 12:56:01.436157046 -0500
@@ -342,7 +342,7 @@
             if self.target_triple.is_empty() {
                 return Err("Unknown non-native platform".to_string());
             }
-            linker = Some(format!("-Clinker={}-gcc", self.target_triple));
+            linker = Some(format!("-Clinker=sh-link-wrapper"));
             self.run_in_vm = true;
         }
 
